File: src/graphControler.ml
(* 
 * Copyright (C) 2015 Gaëtan Dubreil
 *
 *  All rights reserved.This file is distributed under the terms of the
 *  GNU General Public License version 3.0.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *)

open Graffophone_plugin
open Usual

module Bus = EventBus
module Tkr = Talker

class c = object (self)

  val mutable mSelectedEar : (Tkr.c * int) option = None
  val mutable mSelectedVoice : (Tkr.c * int) option = None
  val mutable mNewTalker : Tkr.c option = None
  val mutable mSelectedTalkers : Tkr.c list = []
  val mutable mControlKeyPressed = false
  val mutable mShiftKeyPressed = false
  val mutable mAltKeyPressed = false


  method setControlKeyPressed v = mControlKeyPressed <- v
  method setShiftKeyPressed v = mShiftKeyPressed <- v
  method setAltKeyPressed v = mAltKeyPressed <- v

  method setSelectedTalker tkr =

    if mControlKeyPressed then (

      if L.memq tkr mSelectedTalkers then (
        mSelectedTalkers <- L.filter (fun t -> t != tkr) mSelectedTalkers;
        Bus.notify(Bus.TalkerUnselected tkr#getId)
      )
      else (
        mSelectedTalkers <- tkr::mSelectedTalkers;
        Bus.notify(Bus.TalkerSelected tkr#getId);
      )
    )
    else (
      if L.length mSelectedTalkers > 1 then (
        L.iter(fun tkr -> Bus.notify(Bus.TalkerUnselected tkr#getId)) mSelectedTalkers;
        mSelectedTalkers <- [tkr];
        Bus.notify(Bus.TalkerSelected tkr#getId);
      )
      else (
        if L.memq tkr mSelectedTalkers then (
          mSelectedTalkers <- [];
          Bus.notify(Bus.TalkerUnselected tkr#getId)
        )
        else (
          mSelectedTalkers <- [tkr];
          Bus.notify(Bus.TalkerSelected tkr#getId);
        )
      )
    )


  method setTalkerName (tkr:Tkr.c) v =
    tkr#setName v;
    Bus.notify Bus.TalkerChanged;


  method setTalkerValue (tkr:Tkr.c) v fly =
    tkr#setValue v;
    if not fly then Bus.notify Bus.TalkerChanged;


  method setTalkerEarToValueByIndex (tkr:Tkr.c) index value fly =
    tkr#setEarToValueByIndex index value;
    if not fly then Bus.notify Bus.TalkerChanged;


  method addTalkerEarToValueByIndex (tkr:Tkr.c) index value =
    tkr#addEarToValueByIndex index value;
    Bus.notify Bus.TalkerChanged;


  method selectEar talker index =
    match mSelectedEar with
    | Some (tkr, idx) -> (
        if talker != tkr || index <> idx then (
          mSelectedEar <- Some (talker, index);
          Bus.notify(Bus.EarSelected (talker#getId, index))
        )
        else (
          mSelectedEar <- None
        );
        Bus.notify(Bus.EarUnselected (tkr#getId, idx));
      )
    | None -> (
        match mSelectedVoice with
        | None ->
          mSelectedEar <- Some (talker, index);
          Bus.notify(Bus.EarSelected (talker#getId, index))
        | Some (tkr, idx) ->
          if tkr != talker then (
            let voice = tkr#getVoices.(idx) in
            talker#setEarToVoiceByIndex index voice;
            mSelectedVoice <- None;
            Bus.notify(Bus.VoiceUnselected (tkr#getId, idx));
            Bus.notify Bus.TalkerChanged;
          )
      )


  method selectVoice talker index =
    match mSelectedVoice with
    | Some (tkr, idx) -> (
        if talker != tkr || index <> idx then (
          mSelectedVoice <- Some (talker, index);
          Bus.notify(Bus.VoiceSelected (talker#getId, index))
        )
        else (
          mSelectedVoice <- None
        );
        Bus.notify(Bus.VoiceUnselected (tkr#getId, idx));
      )
    | None -> (
        match mSelectedEar with
        | None ->
          mSelectedVoice <- Some (talker, index);
          Bus.notify(Bus.VoiceSelected (talker#getId, index))
        | Some (tkr, idx) ->
          if tkr != talker then (
            let voice = talker#getVoices.(index) in
            tkr#setEarToVoiceByIndex idx voice;
            mSelectedEar <- None;
            Bus.notify(Bus.EarUnselected (tkr#getId, idx));
            Bus.notify Bus.TalkerChanged;
          )
      )


  method addEar (talker:Tkr.c) (rootIndex:int) =

    match mSelectedVoice with None -> ()
                            | Some (tkr, idx) ->
                              talker#addEarToVoiceByIndex rootIndex (tkr#getVoices.(idx));
                              Bus.notify Bus.TalkerChanged


  method supEar (talker:Tkr.c) (index:int) =
    talker#supEarByIndex index;
    Bus.notify Bus.TalkerChanged;


  method addNewTalker talker =

    ignore(match mSelectedVoice with None -> ()
                                   | Some (tkr, idx) -> Bus.notify(Bus.VoiceUnselected (tkr#getId, idx)));

    ignore(match mSelectedEar with None -> ()
                                 | Some (tkr, idx) -> Bus.notify(Bus.EarUnselected (tkr#getId, idx)));

    ignore(match mNewTalker with None -> ()
                               | Some tkr ->
                                 mSelectedVoice <- Some (tkr, 0);
                                 Bus.notify(Bus.VoiceSelected (tkr#getId, 0))
      );

    mNewTalker <- Some talker;


  method getNewTalker = mNewTalker

  method deleteTalker (talker:Tkr.c) = trace("Delete talker "^ talker#getName);
    Session.supTalker talker;
    Bus.notify Bus.TalkerChanged;


  method selectTalk (talk:Tkr.talk_t) =
    let tkr = Ear.getTalkTalker talk in
    let port = Ear.getTalkPort talk in
    Bus.notify (Bus.TalkSelected(tkr#getId, port));

end

File: src/graph_controler.rs
use std::cell::RefCell;
use std::collections::HashSet;
use std::rc::Rc;

use talker::identifier::Identifiable;
use talker::identifier::{Id, Index};
use talker::talker::{RTalker, Talker, TalkerBase};

use crate::session_presenter::RSessionPresenter;
use session::event_bus::{Notification, REventBus};

pub struct GraphControler {
    selected_ear: Option<(RTalker, i32)>,
    selected_voice: Option<(RTalker, i32)>,
    new_talker: Option<RTalker>,
    selected_talkers: HashSet<Id>,
    control_key_pressed: bool,
    shift_key_pressed: bool,
    alt_key_pressed: bool,
    session_presenter: RSessionPresenter,
}

pub type RGraphControler = Rc<RefCell<GraphControler>>;

impl GraphControler {
    pub fn new(session_presenter: RSessionPresenter) -> GraphControler {
        Self {
            selected_ear: None,
            selected_voice: None,
            new_talker: None,
            selected_talkers: HashSet::new(),
            control_key_pressed: false,
            shift_key_pressed: false,
            alt_key_pressed: false,
            session_presenter,
        }
    }

    pub fn new_ref(session_presenter: RSessionPresenter) -> RGraphControler {
        Rc::new(RefCell::new(GraphControler::new(session_presenter)))
    }

    pub fn set_control_key_pressed(&mut self, v: bool) {
        self.control_key_pressed = v;
    }
    pub fn set_shift_key_pressed(&mut self, v: bool) {
        self.shift_key_pressed = v;
    }
    pub fn set_alt_key_pressed(&mut self, v: bool) {
        self.alt_key_pressed = v;
    }

    pub fn set_selected_talker(&mut self, tkr: &RTalker) {
        let tkr_id = tkr.borrow().id();

        if self.control_key_pressed || self.selected_talkers.len() < 2 {
            if self.selected_talkers.contains(&tkr_id) {
                self.selected_talkers.remove(&tkr_id);
                self.session_presenter
                    .borrow()
                    .notify(Notification::TalkerUnselected(tkr_id));
            } else {
                self.selected_talkers.insert(tkr_id);
                self.session_presenter
                    .borrow()
                    .notify(Notification::TalkerSelected(tkr_id));
            }
        } else {
            for id in &self.selected_talkers {
                self.session_presenter
                    .borrow()
                    .notify(Notification::TalkerUnselected(*id));
            }
            self.selected_talkers.clear();
            self.selected_talkers.insert(tkr_id);
            self.session_presenter
                .borrow()
                .notify(Notification::TalkerSelected(tkr_id));
        }
    }

    pub fn set_talker_name(&mut self, tkr: &RTalker, v: &str) {
        tkr.borrow_mut().set_name(v);

        self.session_presenter
            .borrow()
            .notify(Notification::TalkerRenamed(tkr.borrow().id()));
    }

    pub fn set_talker_data(&mut self, tkr: &RTalker, v: &str, fly: bool) {
        match tkr.borrow_mut().set_data_from_string(v) {
            Ok(state) => {
                if !fly {
                    self.session_presenter
                        .borrow()
                        .notify(Notification::TalkerChanged);
                }
            }
            Err(e) => self.session_presenter.borrow().notify_error(e),
        }
    }

    pub fn set_talker_ear_value_by_index(
        &mut self,
        tkr: &RTalker,
        ear_idx: usize,
        talk_idx: usize,
        value: f32,
        fly: bool,
    ) {
        match tkr
            .borrow()
            .set_ear_value_by_index(ear_idx, talk_idx, value)
        {
            Ok(state) => {
                if !fly {
                    self.session_presenter
                        .borrow()
                        .notify(Notification::TalkerChanged);
                }
            }
            Err(e) => self.session_presenter.borrow().notify_error(e),
        }
    }

    pub fn add_talker_ear_value_by_index(&mut self, tkr: &RTalker, ear_idx: usize, value: f32) {
        match tkr.borrow().add_ear_value_by_index(ear_idx, value) {
            Ok(state) => {
                self.session_presenter
                    .borrow()
                    .notify(Notification::TalkerChanged);
            }
            Err(e) => self.session_presenter.borrow().notify_error(e),
        }
    }

    /*
    pub fn  selectEar talker index =
      match self.selected_ear with
      | Some (tkr, idx) -> {
          if talker != tkr || index <> idx {
            self.selected_ear = Some (talker, index);
            Bus.notify(Bus.EarSelected (talker#getId, index))
          }
          else {
            self.selected_ear = None
          };
          Bus.notify(Bus.EarUnselected (tkr#getId, idx));
        }
      | None -> {
          match self.selected_voice with
          | None ->
            self.selected_ear = Some (talker, index);
            Bus.notify(Bus.EarSelected (talker#getId, index))
          | Some (tkr, idx) ->
            if tkr != talker {
              let voice = tkr#getVoices.(idx) in
              talker#setEarToVoiceByIndex index voice;
              self.selected_voice = None;
              Bus.notify(Bus.VoiceUnselected (tkr#getId, idx));
              Bus.notify Bus.TalkerChanged;
            }
        }


    pub fn  selectVoice talker index =
      match self.selected_voice with
      | Some (tkr, idx) -> {
          if talker != tkr || index <> idx {
            self.selected_voice = Some (talker, index);
            Bus.notify(Bus.VoiceSelected (talker#getId, index))
          }
          else {
            self.selected_voice = None
          };
          Bus.notify(Bus.VoiceUnselected (tkr#getId, idx));
        }
      | None -> {
          match self.selected_ear with
          | None ->
            self.selected_voice = Some (talker, index);
            Bus.notify(Bus.VoiceSelected (talker#getId, index))
          | Some (tkr, idx) ->
            if tkr != talker {
              let voice = talker#getVoices.(index) in
              tkr#setEarToVoiceByIndex idx voice;
              self.selected_ear = None;
              Bus.notify(Bus.EarUnselected (tkr#getId, idx));
              Bus.notify Bus.TalkerChanged;
            }
        }


    pub fn  showVoice (talker:Tkr.c) port){trace("showVoice tkr "^soi talker#getId^" port "^soi port);
      Bus.notify (Bus.TalkSelected(talker#getId, port));


    pub fn  addEar (talker:Tkr.c) (rootIndex:int) =

      match self.selected_voice with
        None -> ()
      | Some (tkr, idx) ->
        talker#addEarToVoiceByIndex rootIndex (tkr#getVoices.(idx));
        Bus.notify Bus.TalkerChanged


    pub fn  supEar (talker:Tkr.c) (index:int) =
      talker#supEarByIndex index;
      Bus.notify Bus.TalkerChanged;


    pub fn  addNewTalker talker =

      ignore(match self.selected_voice
             with None -> ()
                | Some (tkr, idx) -> Bus.notify(Bus.VoiceUnselected (tkr#getId, idx)));

      ignore(match self.selected_ear
             with None -> ()
                | Some (tkr, idx) -> Bus.notify(Bus.EarUnselected (tkr#getId, idx)));

      ignore(match self.new_talker with
            None -> ()
          | Some tkr ->
            self.selected_voice = Some (tkr, 0);
            Bus.notify(Bus.VoiceSelected (tkr#getId, 0))
        };

      self.new_talker = Some talker;


    pub fn  getNewTalker){self.new_talker

    pub fn  deleteTalker (talker:Tkr.c)){trace("Delete talker "^ talker#getName);
      Session.supTalker talker;
      Bus.notify Bus.TalkerChanged;

                                         */
}
